compiler/rustc_resolve/src/ident.rs

```diff
@@ -962,26 +987,29 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         // Check if one of single imports can still define the name,
         // if it can then our result is not determined and can be invalidated.
         for single_import in &resolution.single_imports {
-            let Some(import_vis) = single_import.vis.get() else {
-                // This branch handles a cycle in single imports, which occurs
-                // when we've previously **steal** the `vis` value during an import
-                // process.
+            let Some(import_vis) = single_import.vis else {
+                continue;
+            };
+            if !self.is_accessible_from(import_vis, parent_scope.module) {
```

Comment:
It's better to move this check to its old place, the ignored import check is cheaper.